### PR TITLE
[MOD-5575] [GeoShape] Geometries validation and correction

### PR DESCRIPTION
**Pull request by ephraimfeldblum**
[MOD-5575](https://redislabs.atlassian.net/browse/MOD-5575) Fixes and optimizations

[MOD-5575]: https://redislabs.atlassian.net/browse/MOD-5575?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/geometry/allocator/allocator.hpp [hunk 1]
```diff
@@ -14,28 +14,28 @@ template <class T>
...
 
-  explicit inline Allocator() = default;
+  explicit inline constexpr Allocator() = default;
   template <class U>
-  explicit inline Allocator(Allocator<U> const&) noexcept;
+  explicit inline constexpr Allocator(Allocator<U> const&) noexcept;
 
-  [[nodiscard]] inline auto allocate(size_t n) noexcept -> value_type*;
-  inline void deallocate(value_type* p, size_t n) noexcept;
+  [[nodiscard]] static inline auto allocate(std::size_t n) noexcept -> value_type*;
```
Comment (oshadmi) [inline src/geometry/allocator/allocator.hpp:21]: Failed on Mac on other PR due to  `std`. Maybe an include is missing? (cstddef)

File: src/geometry/allocator/stateful_allocator.hpp [hunk 2]
```diff
@@ -7,47 +7,47 @@
...
+  inline void deallocate(value_type* p, std::size_t n) noexcept;
 
-  [[nodiscard]] inline constexpr size_t report() const noexcept;
+  [[nodiscard]] inline constexpr std::size_t report() const noexcept;
 };
 
 template <class T>
 template <class U>
-inline StatefulAllocator<T>::StatefulAllocator(StatefulAllocator<U> const&) noexcept {
+inline constexpr StatefulAllocator<T>::StatefulAllocator(StatefulAllocator<U> const&) noexcept {
```
Comment (oshadmi) [inline src/geometry/allocator/stateful_allocator.hpp:31]: Add a **short** description on the class (what is it good for etc.)

File: src/geometry/rtree.hpp [hunk 3]
```diff
@@ -36,68 +31,61 @@ namespace bgi = bg::index;
...
+  explicit RTree();
 
-  [[nodiscard]] auto from_wkt(std::string_view wkt) const -> geom_type;
-  void insert(geom_type const& geom, t_docId id);
   int insertWKT(std::string_view wkt, t_docId id, RedisModuleString** err_msg);
   bool remove(t_docId id);
+  [[nodiscard]] auto query(std::string_view wkt, QueryType query_type,
+                           RedisModuleString** err_msg) const -> IndexIterator*;
 
-  [[nodiscard]] static auto geometry_to_string(geom_type const& geom) -> string;
```
Comment (oshadmi) [inline src/geometry/rtree.hpp:79]: Consider moving to another file for doc services or somthing like this

File: tests/pytests/test_geometry_flat.py [hunk 4]
```diff
@@ -264,7 +264,7 @@ def testFtInfo(env):
   res = to_dict(env.cmd('FT.INFO idx1'))
   env.assertEqual(float(res[info_key_name]), cur_usage)
 
-  doc_num = 100
+  doc_num = 10000
```
Comment (oshadmi) [inline tests/pytests/test_geometry_flat.py:267]: Is this enough to cause mem relocation? Can we somehow make sure it does?

File: src/geometry/rtree.hpp (hunk 3 above)
Comment (ephraimfeldblum) [reply src/geometry/rtree.hpp:79]: Next PR.

File: tests/pytests/test_geometry_flat.py (hunk 4 above)
Comment (ephraimfeldblum) [reply tests/pytests/test_geometry_flat.py:267]: Experimentally. This value always caused a mem relocation related crash prior to this fix.

We should, however, find a way to force mem relocation regardless.

**Review by oshadmi — APPROVED** [review]

**Pull request merged**